javaslang/src/main/java/javaslang/collection/JavaConverters.java

```diff
@@ -0,0 +1,561 @@
+ * <li>javaslang.collection.Map has a java.util.Map view</li>
+ * <li>javaslang.collection.Multimap has a java.util.Map view</li>
+ * </ul>
+ *
+ * Subtypes of the Javaslang types mentioned above can have special views that make use of optimized implementations.
+ *
+ * @author Pap Lőrinc, Daniel Dietrich
+ * @since 2.1.0
+ */
+final class JavaConverters {
+    private JavaConverters() { }
+
+    static <T> java.util.ListIterator<T> listIterator(IndexedSeq<T> seq, int index) {
+        return new IndexedSeqListIterator<>(seq, index);
+    }
+
+    static <T> java.util.List<T> asJava(LinearSeq<T> seq) {
+        return new SeqAsJavaList<>(seq);
+    }
+
+    static <T> java.util.List<T> asJava(IndexedSeq<T> seq) {
+        return new SeqAsJavaList<T>(seq) {
+            @Override
+            public java.util.ListIterator<T> listIterator() {
+                return JavaConverters.listIterator(((IndexedSeq<T>) delegate), 0);
+            }
+
+            @Override
+            public java.util.ListIterator<T> listIterator(int index) {
+                return JavaConverters.listIterator(((IndexedSeq<T>) delegate), index);
+            }
+        };
+    }
+
+    static <T> java.util.Set<T> asJava(Set<T> set) {
+        return new SetAsJavaSet<>(set);
+    }
+
+    static <K, V> java.util.Map<K, V> asJava(Map<K, V> map) {
+        return new MapAsJavaMap<>(map);
+    }
+
+    static <K, V> java.util.Map<K, java.util.Collection<V>> asJava(Multimap<K, V> map) {
+        return new MultimapAsJavaMap<>(map);
+    }
+
+    // -- private view implementations
+
+    private static class SeqAsJavaList<T> implements java.util.List<T>, Serializable {
+        private static final long serialVersionUID = 1L;
+        protected Seq<T> delegate;
+
+        SeqAsJavaList(Seq<T> delegate) {
+            this.delegate = delegate;
+        }
+
+        @Override
+        public boolean add(T element) {
+            delegate = delegate.append(element);
+            return true;
+        }
+
+        @Override
+        public void add(int index, T element) {
+            // may throw an IndexOutOfBoundsException accordingly to the j.u.List.add(int, T)
+            delegate = delegate.insert(index, element);
+        }
+
+        @Override
+        public boolean addAll(Collection<? extends T> collection) {
+            delegate = delegate.appendAll(collection);
+            return true;
+        }
+
+        @Override
+        public boolean addAll(int index, Collection<? extends T> collection) {
+            // may throw an IndexOutOfBoundsException accordingly to the j.u.List.addAll(int, Collection)
+            delegate = delegate.insertAll(index, collection);
+            return true;
+        }
+
+        @Override
+        public void clear() {
+            delegate = delegate.take(0);
+        }
+
+        @Override
+        public boolean contains(Object obj) {
+            return Seq.<Object>narrow(delegate).contains(obj);
```

Comment:
(Note: In this case we prefer composition over inheritance to add specific behavior. That's the reason I added the ListIteratorFactory)